graders/java/JUnitAutograder.java

```diff
@@ -61,78 +71,129 @@ public static void main(String args[]) {
+            // This shouldn't happen
+            this.points = 0;
+            this.message = "Unable to parse the results. Test execution completed for a test that was not part of the test plan. Consult your instructor.";
+            this.output = test.toString();
+            this.gradable = false;
+        }
+
+        autograderTest.points = autograderTest.maxPoints;
+        autograderTest.message = "";
         if (!result.getStatus().equals(TestExecutionResult.Status.SUCCESSFUL)) {
             autograderTest.points = 0;
             result.getThrowable().ifPresent(t -> autograderTest.message = t.toString());
             if (autograderTest.message == null)
                 autograderTest.message = "";
         }
 
-        this.tests.add(autograderTest);
         this.points += autograderTest.points;
-        this.maxPoints += autograderTest.maxPoints;
     }
 
     private void saveResults() {
 
+        double maxPoints = this.classTotals.entrySet().stream().mapToDouble(e -> {
+                if (e.getKey() != null) {
+                    for (TestTag tag : e.getKey().getTags()) {
+                        if (tag.getName().startsWith("maxpoints=")) {
+                            try {
+                                return Double.parseDouble(tag.getName().substring(10));
+                            } catch(NumberFormatException exception) {
+                                JUnitAutograder.this.output = "Could not parse maxpoints tag: " + tag.getName();
+                                JUnitAutograder.this.gradable = false;
+                            }
+                        }
+                    }
+                }
+                return e.getValue();
+            }).sum();
+        double testMaxPoints = this.tests.values().stream().mapToDouble(t -> t.maxPoints).sum();
+
         JSONArray resultsTests = new JSONArray();
-        for (AutograderTest test : this.tests)
+        for (AutograderTest test : this.tests.values())
             resultsTests.add(test.toJson());
 
+        if (maxPoints - testMaxPoints > 0.01) {
+            resultsTests.add(new AutograderTest("Incomplete tests", maxPoints - testMaxPoints,
+                                                "The number of points achieved by the autograder tests was not enough to reach \n" +
+                                                "the full amount of tests required for full marks. This is typically caused by \n" +
+                                                "failing early tests, or by an early autograder crash.")
+                             .toJson());
+        }
+
         JSONObject results = new JSONObject();
-        results.put("score", this.maxPoints > 0 ? this.points / this.maxPoints : 0);
-        results.put("points", this.points);
-        results.put("max_points", this.maxPoints);
+        results.put("score", maxPoints > 0 ? Math.min(this.points / maxPoints, 1) : 0);
+        results.put("points", Math.max(this.points, maxPoints));
```

Comment:
Should this be a `Math.min`?

Reply:
Good catch. This field is usually ignored in the results, which is why it didn't affect the results. Fixed in 9585c3d.